scripts/build_test.py

```diff
@@ -14,29 +14,313 @@
+        """Test process_html removes whitespaces and adds hash to filepaths."""
+        BASE_HTML_SOURCE_PATH = os.path.join(TEMPLATES_DEV_DIR, 'base.html')
+        BASE_JS_SOURCE_PATH = os.path.join(
+            TEMPLATES_DEV_DIR, 'pages', 'Base.js')
+        build._ensure_files_exist([BASE_HTML_SOURCE_PATH, BASE_JS_SOURCE_PATH])
+        # Prepare a file_stream object from StringIO.
+        minified_html_file_stream = StringIO.StringIO()
+        # Obtain actual file hashes of /templates to add hash to all filepaths
+        # within the HTML file. The end result will look like:
+        # E.g <script ... app.js></script>
+        # --> <script ... app.[hash].js></script>.
+        # Only need to hash Base.js.
+        with self.swap(build, 'FILE_EXTENSIONS_TO_IGNORE', ('.html',)):
+            file_hashes = build.get_file_hashes(TEMPLATES_DEV_DIR)
+
+        # Assert that base.html has white spaces and has original filepaths.
+        with open(BASE_HTML_SOURCE_PATH, 'r') as source_base_file:
+            source_base_file_content = source_base_file.read()
+            self.assertRegexpMatches(
+                source_base_file_content, r'\s{2,}',
+                msg="No white spaces detected in %s unexpectedly"
+                % BASE_HTML_SOURCE_PATH)
+            for filepath in file_hashes:
+                # Looking for templates/pages/Base.js in
+                # source_base_file_content.
+                self.assertTrue(
+                    re.search(filepath, source_base_file_content))
+
+        # Build base.html file.
+        with open(BASE_HTML_SOURCE_PATH, 'r') as source_base_file:
+            build.process_html(
+                source_base_file, minified_html_file_stream, file_hashes)
+
+        minified_html_file_content = minified_html_file_stream.getvalue()
+        self.assertNotRegexpMatches(
+            minified_html_file_content, r'\s{2,}',
+            msg='All white spaces must be removed from %s' %
+            BASE_HTML_SOURCE_PATH)
+        # Assert that hash is inserted into filenames in base.html.
+        # Final filepath in base.html example:
+        # /build/templates/head/pages/Base.081ce90f17ecdf07701d83cb860985c2.js.
+        for filepath in file_hashes:
+            final_filename = build._insert_hash(filepath, file_hashes[filepath])
+            # Looking for
+            # templates/pages/Base.081ce90f17ecdf07701d83cb860985c2.js in
+            # minified_html_file_content.
+            self.assertTrue(
+                re.search(final_filename, minified_html_file_content))
+
+    def test_file_should_be_built(self):
+        """Test file_should_be_built returns the correct boolean value for
+        filepath that should be built.
+        """
+        service_js_filepath = os.path.join('core', 'pages', 'AudioService.js')
+        spec_js_filepath = os.path.join('core', 'pages', 'AudioServiceSpec.js')
+        # At the point of writing this, Protractor files in /extensions share
```

Comment:
I don't understand this comment (or the reason for it) -- could you please elaborate?

Reply:
I added comment as a note to reader should they wonder why their `protractor.js` file is not served in production.

`should_file_be_built` filters extraneous files like Karma tests, Python tests files...etc. There are a bunch of Protractor files for e2e testings named `protractor.js` within `/extensions/interactions`, that I am aware of. 

In the future, there could be Protractor files, that is not named `protractor.js`, get added to `/extensions/` so looking for JS files with the name :`protractor.js` would suffice. I think this is the 2nd purpose of this comment.

Should someone is revising build.py with regards to Protractor files in production build, they can relax / restrict the filename rule as well.